share/availability/full/reconstruction_test.go

```diff
@@ -26,49 +27,99 @@ func init() {
 // light nodes only.
 func TestShareAvailable_OneFullNode(t *testing.T) {
 	// NOTE: Numbers are taken from the original 'Fraud and Data Availability Proofs' paper
-	light.DefaultSampleAmount = 20 // s
-	const (
-		origSquareSize = 16 // k
-		lightNodes     = 69 // c
-	)
-
-	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
-	defer cancel()
-
-	net := availability_test.NewTestDAGNet(ctx, t)
-	source, root := RandNode(net, origSquareSize) // make a source node, a.k.a bridge
-	full := Node(net)                             // make a full availability service which reconstructs data
-
-	// ensure there is no connection between source and full nodes
-	// so that full reconstructs from the light nodes only
-	net.Disconnect(source.ID(), full.ID())
-
-	errg, errCtx := errgroup.WithContext(ctx)
-	errg.Go(func() error {
-		return full.SharesAvailable(errCtx, root)
-	})
-
-	lights := make([]*availability_test.TestNode, lightNodes)
-	for i := 0; i < len(lights); i++ {
-		lights[i] = light.Node(net)
-		go func(i int) {
-			err := lights[i].SharesAvailable(ctx, root)
-			if err != nil {
-				t.Log("light errors:", err)
-			}
-		}(i)
+	tc := []struct {
+		name            string
+		origSquareSize  int  // k
+		lightNodes      int  // c
+		sampleAmount    uint // s
+		recoverability  Recoverability
+		expectedFailure bool
+	}{
+		{
+			name:            "fully recoverable",
+			origSquareSize:  16,
+			lightNodes:      24, // ~99% chance of recoverability
+			sampleAmount:    20,
+			recoverability:  FullyRecoverable,
+			expectedFailure: false,
+		},
+		{
+			name:            "fully recoverable, not enough LNs",
+			origSquareSize:  16,
+			lightNodes:      19, // ~0.7% chance of recoverability
+			sampleAmount:    20,
+			recoverability:  FullyRecoverable,
+			expectedFailure: true,
+		},
+		{
+			name:            "barely recoverable",
+			origSquareSize:  16,
+			lightNodes:      230, // 99% chance of recoverability
```

Comment:
Again, this case is not covered in the paper. Refer to my description, and I can send you the script so you can verify it for yourself

Reply:
The description didn't describe where 230 is coming from and thus the question. 

> Again, this case is not covered in the paper
Okay, then I missed some discussion around it because I don't remember such a fact. In my memory, figures from the paper assume withholding attack as it is extensively described in theorems in the same sections as all the plots I am referring to. 

Please share the script, I am curious about how it's constructed and would like to play and verify with it. The number 230 feels too much for me when 23 is enough to reconstruct the whole square. Did you verify the script with @musalbas

Reply:
I meant 23/24 when data is not withheld, update the comment.

Reply:
https://gist.github.com/distractedm1nd/d2b306da13581b97bbbcefaa5b050738
I apologize for the shitty code, didnt expect it to be read by others :)

Reply:
Main loop flow looks familiar 🙂  
If `ds.sample(s)` is acts as noop (for example if share is withheld), it would mean that given share is not sampled, but you count it as sampled anyway here:
```
for len(uniqueSamples) < shares {
	s := randomSample(k)
	uniqueSamples[s]++
	ds.sample(s)
}
```
Could you please clarify, why you count it is not sampled? Looks like this code could lead to less than 20 samples being assigned to each light node and make required amount of light nodes higher that it needs to be. 
It seems like code should be as following:
```
for len(uniqueSamples) < shares {
	s := randomSample(k)
	if ds.sample(s) {
		uniqueSamples[s]++
	}
}
```
				
On adjusted calculation algo, probability graph shows, that there is decent chance to do reconstruction for withholding case on amount of light node lower than 230:
<img width="612" alt="image" src="https://github.com/celestiaorg/celestia-node/assets/13818348/c4aba40b-3c53-46f6-bc9f-4523b5189f96">

Reply:
Yes - I tried to match it with the current implementation that does not change the coordinates to sample after failure but just retries. Or is this incorrect?

Reply:
I've checked light availability and seems like it is acting like neither of simulations. 😆  I'll drop my finding into new issue to discuss further.

Reply:
https://github.com/celestiaorg/celestia-node/issues/2780